itm/src/iter.rs

```diff
@@ -0,0 +1,788 @@
+                            malformed_packets,
+                            consumed_packets,
+                        });
+                    }
+
+                    // A global timestamp: store until we have both the
+                    // upper (GTS2) and lower (GTS1) bits.
+                    TracePacket::GlobalTimestamp1 { ts, wrap, clkch } => {
+                        gts.replace_lower(ts);
+
+                        if wrap {
+                            // upper bits have changed; GTS2 incoming
+                            gts.upper = None;
+                        } else {
+                            apply_gts(&gts, &mut self.current_baseline, &options);
+                        }
+                        if clkch {
+                            // system has asserted clock change input; full GTS incoming
+                            //
+                            // A clock change signal that the system
+                            // asserts if there is a change in the ratio
+                            // between the global timestamp clock
+                            // frequency and the processor clock
+                            // frequency. Implementation and use of the
+                            // clock change signal is optional and
+                            // deprecated.
+                            gts.reset();
+                        }
+                    }
+                    TracePacket::GlobalTimestamp2 { ts } => {
+                        gts.upper = Some(ts);
+                        apply_gts(&gts, &mut self.current_baseline, &options);
+                    }
+
+                    packet => packets.push(packet),
+                },
+            }
+        }
+    }
+}
+
+impl<'a, R> Iterator for Timestamps<'a, R>
+where
+    R: Read,
+{
+    type Item = Result<TimestampedTracePackets, DecoderError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let trace = self.next_timestamped(self.options.clone());
+
+        match trace {
+            Err(DecoderErrorInt::Eof) => None,
+            Err(DecoderErrorInt::Io(io)) => Some(Err(DecoderError::Io(io))),
+            Err(DecoderErrorInt::MalformedPacket(m)) => Some(Err(DecoderError::MalformedPacket(m))),
+            Ok(trace) => Some(Ok(trace)),
+        }
+    }
+}
+
+fn calc_offset(ts: u64, prescaler: Option<LocalTimestampOptions>, freq: u32) -> chrono::Duration {
+    let prescale = match prescaler {
+        None | Some(LocalTimestampOptions::Enabled) => 1,
+        Some(LocalTimestampOptions::EnabledDiv4) => 4,
+        Some(LocalTimestampOptions::EnabledDiv16) => 16,
+        Some(LocalTimestampOptions::EnabledDiv64) => 64,
+        Some(LocalTimestampOptions::Disabled) => unreachable!(), // checked in `Timestamps::new`
+    };
+    let ticks = ts * prescale;
+    let seconds = (1_f64 / freq as f64) * ticks as f64;
+
+    // NOTE(ceil) ...
+    chrono::Duration::nanoseconds((seconds * 1e9).ceil() as i64)
```

Comment:
By rounding up timestamps they will consistently always be at most 1ns too late, rather than a packet occuring 1ns *before* the hardware event, or being inconsistent if the timestamp was rounded up or down depending on the value.